LEBCs implementation

wkdarko opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #???

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Checklist:

- [ ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [ ] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [ ] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [ ] I have summarized these changes in `CHANGELOG.rst` following the established format.

wkdarko commented:
Hello @mphoward, the first draft of this PR is ready for your preview please. In this, I have used the modified wrap method of the Box in, for example, the TwoStepConstantVolume class. However, there are a few other files where the original wrap method is being used also, and I was wondering whether the modified wrap should be used across the board or in only selected places of relevance.

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -387,6 +387,116 @@ struct
+        Scalar3 dv;
+    };
+
+    HOSTDEVICE MinImage minImage(const Scalar3& p, const Scalar3& v) const
```
Rather than make a struct, put both as arguments by reference:
```suggestion
    HOSTDEVICE void minImage(Scalar3& dx, Scalar3& dv, const Scalar3& p, const Scalar3& v) const
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -387,6 +387,116 @@ struct
+        if (m_periodic.z)
+            {
+            Scalar img = slow::rint(pos.z * m_Linv.z);
+            auto sign = std::copysign(Scalar(1.0), img);
```
Is this necessary? `img` is supposed to always be -1, 0, and 1 (see doxygen comment), so you should be able to update `vel` like is done for `pos`. I think you want a `-=`, but confirm with your equations.

Just FYI: use of the STL is also discouraged for code that needs to work with CUDA if not required, as not all the STL is supported in CUDA.

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -387,6 +387,116 @@ struct
+            }
+#endif
+
+        return {pos, vel};
```
Once this is void, no return will be needed.
```suggestion
```

mphoward commented on hoomd/extern/hipCUB:
Your git submodules seem to have gotten messed up if they are included in a commit. I think they'll be fine for now, but we'll need to revert these to what's on `trunk-minor` later. I can help you figure that out when the time comes

wkdarko commented:
Hi @mphoward, could you please give my PR another round of review? Thank you.

In the latest commit, I have applied the new wrapping method to all classes that use "wrap". For the modified "minImage" method, I could find only a few instances where the new change might be relevant (one in the "ForceDistanceConstraint" class and the other in "PotentialPairDPDThermo").

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +393,120 @@ struct
+        if (m_periodic.x)
+            {
+            r.x -= L.x * slow::rint(r.x * m_Linv.x);
+            v.x -= m_L_rate.x * slow::rint(r.x * m_Linv.x);
```
Store intermediate result to avoid calling same function twice:
```suggestion
            Scalar img = slow::rint(r.x * m_Linv.x);
            r.x -= L.x * img;
            v.x -= m_L_rate.x * img;
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -537,6 +770,37 @@ struct
+    
+        r = pos;
+        v = vel;
+        }
```
Two things here:
* Renaming the variables is a good idea to make things clearer, but it should be done at the argument level.
* You can call the base method safely here to avoid code duplication.

```suggestion
    //! Shift a particle by a multiple of the lattice vectors
    /*! \param pos The position vector to shift (and the new shifted position)
        \param vel The velocity vector to shift (and the new shifted velocity)
        \param _shift The displacement in lattice coordinates
     */
    HOSTDEVICE void shift(Scalar3& pos, Scalar3& vel, const int3& _shift) const
        {
        shift(pos, _shift);        
        vel += Scalar(_shift.x) * make_scalar3(m_L_rate.x, 0.0, 0.0);
        vel += Scalar(_shift.y) * make_scalar3(m_L.y * m_xy_rate, m_L_rate.y, 0.0);
        vel += Scalar(_shift.z) * make_scalar3(m_L.z * m_xz_rate, m_L.z * m_yz_rate, m_L_rate.z);
        }
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -537,6 +770,37 @@ struct
+
+    //! Shift a vec3
+    HOSTDEVICE void shift(vec3<Scalar>& r, vec3<Scalar>& v, const int3& _shift) const
+    {
```
This formatting isn't quite right. Make sure you have setup pre-commit in your development environment so it fixes these things for you automatically.

mphoward commented on hoomd/extern/hipCUB:
These changes to the submodules (hipCUB, hipper, and tutorial) need to be reverted. If you are unsure how to do so, let me know and I can try.

mphoward commented on hoomd/md/ForceComposite.cc:
```diff
@@ -579,11 +579,13 @@ void ForceComposite::createRigidBodies(
                 quat<Scalar> constituent_orientation = body_orientation * local_orientation;
 
                 snap.pos[constituent_particle_tag] = constituent_position;
+                snap.vel[constituent_particle_tag] = snap.vel[particle_tag];
```
You should check if you need to merge in a newer version of `trunk-minor` to this branch. Michaela modified some of these functions to set the constituent velocity, so a different change might be required here. Please work with her if you merge that in and something needs to be moved around.

mphoward replied:
A change should also be needed in `updateCompositeParticles`. That method may not have used `wrap` explicitly, but rather `getImage` and `shift` instead.

mphoward commented on hoomd/md/TwoStepConstantPressureGPU.cu:
```diff
@@ -194,28 +195,33 @@ gpu_npt_mtk_wrap_kernel(const unsigned int nwork, Scalar4* d_pos, int3* d_image,
         // fetch particle position
         Scalar4 postype = d_pos[idx];
         Scalar3 pos = make_scalar3(postype.x, postype.y, postype.z);
+        Scalar4 veltype = d_vel[idx];
```
These Scalar4's are actually velocity + mass. Please rename anywhere you have `veltype` for readability.
```suggestion
        Scalar4 velmass = d_vel[idx];
```

mphoward commented on hoomd/md/TwoStepConstantVolumeGPU.cu:
```diff
@@ -77,7 +77,7 @@ __global__ void gpu_nvt_rescale_step_one_kernel(Scalar4* d_pos,
 
         // time to fix the periodic boundary conditions
-        box.wrap(pos, image);
+        box.wrap(pos, vel, image);
```
Please confirm anywhere you're just using the existing velocity variable that you've been given that it is (1) not const and (2) its array handle was not read only. The compiler should yell about the first one but not the second.

mphoward reviewed (commented):
Thanks! This is generally looking very good. Please address these comments, and then it is important that we work on the `shift` cases next.

wkdarko commented:
Hi @mphoward, I would appreciate your feedback on the latest update to my PR. In the recent commit on this PR, I have addressed the comments from the previous review. I have also implemented the `shift` method where it seems to me a change is required, in a few of the 8 classes that use the method. I also added box deformation rates getters and setters to the Box.

A couple of questions that I wondered: 
- In the `ParticleData` class, should the wrap method of the Box be needed in `setVelocity` as used in `setPosition`?
- In the `test/test_pdata.cc` file, assertion tests are performed on the `wrap` and `minImage` methods. Do we perform these tests on just the modified versions of the respective methods, on both the original and modified versions, or leave them as they are on the original versions only?

Thanks.

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+        {
+        Scalar3 r = dr;
+        Scalar3 v = dv;
+        Scalar3 L = getL();
```
```suggestion
        const Scalar3 L = getL();
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+#ifdef __HIPCC__
+        if (m_periodic.z)
+            {
+            Scalar img = slow::rint(r.z * m_Linv.z);
```
```suggestion
            const Scalar img = slow::rint(r.z * m_Linv.z);
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+
+        if (m_periodic.y)
+            {
+            Scalar img = slow::rint(r.y * m_Linv.y);
```
```suggestion
            const Scalar img = slow::rint(r.y * m_Linv.y);
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+
+        if (m_periodic.x)
+            {
+            Scalar img = slow::rint(r.x * m_Linv.x);
```
```suggestion
            const Scalar img = slow::rint(r.x * m_Linv.x);
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+            {
+            if (r.y >= m_hi.y)
+                {
+                int i = int(r.y * m_Linv.y + Scalar(0.5));
```
This is what's called a C-style cast, but this way is preferred now:
```suggestion
                const int i = static_cast<int>(r.y * m_Linv.y + Scalar(0.5));
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+                }
+            else if (r.y < m_lo.y)
+                {
+                int i = int(-r.y * m_Linv.y + Scalar(0.5));
```
```suggestion
                const int i = static_cast<int>(-r.y * m_Linv.y + Scalar(0.5));
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+            {
+            if (r.x >= m_hi.x)
+                {
+                int i = int(r.x * m_Linv.x + Scalar(0.5));
```
```suggestion
                const int i = static_cast<int>(r.x * m_Linv.x + Scalar(0.5));
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -393,6 +452,120 @@ struct
+                }
+            else if (r.x < m_lo.x)
+                {
+                int i = int(-r.x * m_Linv.x + Scalar(0.5));
```
```suggestion
                const int i = static_cast<int>(-r.x * m_Linv.x + Scalar(0.5));
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -496,6 +669,128 @@ struct
+    HOSTDEVICE void
+    wrap(Scalar3& pos, Scalar3& vel, int3& img, char3 flags = make_char3(0, 0, 0)) const
+        {
+        Scalar3 L = getL();
```
```suggestion
        const Scalar3 L = getL();
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -496,6 +669,128 @@ struct
+
+        // tilt factors for nonperiodic boxes are always calculated w.r.t. to the global box with
+        // origin (0,0,0)
+        if (!m_periodic.y)
```
Just wanted to check: there is no need for a check on the x-component here?

wkdarko replied:
No, I do not think so since it is the reference dimension. Also, only the y- and z-components of the origin are used here.

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -496,6 +669,128 @@ struct
+
+        if (m_periodic.x)
+            {
+            Scalar tilt_x = (m_xz - m_xy * m_yz) * (pos.z - origin.z) + m_xy * (pos.y - origin.y);
```
```suggestion
            const Scalar tilt_x = (m_xz - m_xy * m_yz) * (pos.z - origin.z) + m_xy * (pos.y - origin.y);
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -496,6 +669,128 @@ struct
+
+        if (m_periodic.y)
+            {
+            Scalar tilt_y = m_yz * (pos.z - origin.z);
```
```suggestion
            const Scalar tilt_y = m_yz * (pos.z - origin.z);
```

mphoward commented on hoomd/BoxDim.h:
```diff
@@ -496,6 +669,128 @@ struct
+        pos.z = pos_scalar.z;
+        vel.x = vel_scalar.x;
+        vel.y = vel_scalar.y;
+        vel.z = vel_scalar.z;
```
Can you do the same here as in `minImage`?
```suggestion
        pos = vec3<Scalar>(pos_scalar);
        vel = vec3<Scalar>(vel_scalar);
```

mphoward commented on hoomd/ParticleData.cc:
```diff
@@ -1413,13 +1419,17 @@ Scalar3 ParticleData::getPosition(unsigned int tag) const
     unsigned int idx = getRTag(tag);
     bool found = (idx < getN());
     Scalar3 result = make_scalar3(0.0, 0.0, 0.0);
+    Scalar3 vel = make_scalar3(0.0, 0.0, 0.0);
```
Velocity is not going to be needed in this method because it is just returning a position. You should revert the changes here.

mphoward commented on hoomd/ParticleData.cc:
```diff
@@ -1463,10 +1487,17 @@ Scalar3 ParticleData::getVelocity(unsigned int tag) const
+        bcast(pos.z, owner_rank, m_exec_conf->getMPICommunicator());
+        bcast(img.x, owner_rank, m_exec_conf->getMPICommunicator());
+        bcast(img.y, owner_rank, m_exec_conf->getMPICommunicator());
+        bcast(img.z, owner_rank, m_exec_conf->getMPICommunicator());
```
Only the result needs to be broadcast, even though the position & image were needed for the calculation.
```suggestion
```

mphoward commented on hoomd/ParticleData.cc:
```diff
@@ -1463,10 +1487,17 @@ Scalar3 ParticleData::getVelocity(unsigned int tag) const
         }
 #endif
     assert(found);
+    m_global_box->wrap(pos, result, img);
```
Move this calculation inside the `if (found)` branch. Only the wrapped result needs to be broadcast. I'm not sure why the position method didn't do this originally.

mphoward commented on hoomd/ParticleData.cc:
```diff
@@ -1531,7 +1568,7 @@ int3 ParticleData::getImage(unsigned int tag) const
 
     // wrap into correct image
-    m_global_box->wrap(pos, result);
+    m_global_box->wrap(pos, vel, result);
```
Same comments apply here as for velocity, although now you should probably fix the original code: but the shift and the wrap in the `if (found)` branch, then only broadcast the final result.

mphoward commented on hoomd/ParticleData.cc:
```diff
@@ -3225,12 +3271,14 @@ void SnapshotParticleData<Real>::replicate(unsigned int nx,
         // different type
-        p = vec3<Real>(old_box.shift(vec3<Scalar>(p), img));
+        old_box.shift(r, v, img);
+        vec3<Real> p = vec3<Real>(vec3<Scalar>(r));
```
Will this work?
```suggestion
        vec3<Real> p(r);
```

mphoward commented on hoomd/md/test/test_opls_dihedral_force.cc:
I apologize that I misunderstood the question you asked before. These unit tests are never going to set a deformation rate on the box, so I don't think we should add the velocities. Please revert the changes to this file using git. (Lookup the command or ask me if you're unsure.) I will mark the other test files that can also be reverted.

mphoward commented on hoomd/md/test/test_table_angle_force.cc:
Revert

mphoward commented on hoomd/md/test/test_table_dihedral_force.cc:
Revert

mphoward commented on hoomd/test/test_gridshift_correct.cc:
These tests might need to be expanded to include particle velocities and a nonzero deformation rate.

mphoward reviewed (changes requested):
Here are some suggested changes. Please let me know when you have updated the `BoxDim` unit tests, and I'll review those!

wkdarko commented on hoomd/test/test_pdata.cc:
```diff
@@ -415,6 +415,65 @@ UP_TEST(BoxDim_triclinic_test)
+    MY_CHECK_CLOSE(p_old.z, p_new.z, tol);
+
+    // check the wrapped velocities
+    MY_CHECK_CLOSE(vel.x, 1.5, tol);
```
This line fails under `ctest` but passes the test when L_rate.x is added to give 1.7. I do not think this should be the case since the given x-coordinate hasn't crossed its boundary (or has an image of 0).

wkdarko commented:
Hello @mphoward, this PR is ready for one final review if you could please. Thanks.

wkdarko commented on hoomd/test/test_gridshift_correct.cc:
```diff
@@ -116,11 +132,22 @@ UP_TEST(ParticleDataGridShiftGetMethods)
     UP_ASSERT_EQUAL(pimg.y, 0);
     UP_ASSERT_EQUAL(pimg.z, 0);
-    pimg = pdata->getImage(0);
+    pimg = pdata->getImage(1);
```
I think the particle index in the original code should be 1 not 0, so I changed that.
